### PR TITLE
Sonoff: improve device compliance

### PR DESCRIPTION
**Pull request by atrovato**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
~~- [ ] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)~~
~~- [ ] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/gladys-4-docs) and the [website](https://documentation.gladysassistant.com).~~
- [x] Did you add fake requests data for the demo mode (`front/src/config/demo.json`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Complete the device compatibility.
Remove manual creation.
Rename to Tasmota.

### CODE REVIEW

**Comment by atrovato** [comment]
Please @Pierre-Gilles don't forgot this PR, it a great refactor :)

**Comment by Pierre-Gilles** [comment]
@atrovato Don't worry, but the PR is really huge so it'll take a lot of time to review it. 

For now all my time is dedicated to the scene feature, hope you'll understand! 😄

**Comment by atrovato** [comment]
I understand, please don't forget it :)

**Comment by Pierre-Gilles** [comment]
Hi ! I'm reviewing your PR and wanted to see the code coverage, I don't understand why the coverage is never reported on this PR specifically (works with other PR...)

I think it's because tests are passing on your CircleCI, not Gladys Assistant CircleCI ! Any idea ?

File: front/src/components/boxs/device-in-room/device-features/BinaryDeviceFeature.jsx [hunk 1]
```diff
@@ -16,7 +16,9 @@ const BinaryDeviceType = ({ children, ...props }) => {
       <td>
         <i class="fe fe-toggle-right" />
       </td>
-      <td>{props.device.name}</td>
+      <td>
+        {props.device.name} - {props.deviceFeature.name}
```
Comment (Pierre-Gilles) [inline front/src/components/boxs/device-in-room/device-features/BinaryDeviceFeature.jsx]: This is not related to Sonoff, and I don't think we want that ! 

We already had a debate on this subject on the forum ^^ Still open to discussion if you have new arguments about that let's debate (on the forum?)

**Comment by atrovato** [comment]
It don't work all the time, on other PR too.
I try a rebase too force it.

File: server/utils/constants.js [hunk 2]
```diff
@@ -228,6 +228,7 @@ const DEVICE_FEATURE_CATEGORIES = {
   OPENING_SENSOR: 'opening-sensor',
   HUMIDITY_SENSOR: 'humidity-sensor',
   VIBRATION_SENSOR: 'vibration-sensor',
+  COUNTER_SENSOR: 'counter-sensor',
```
Comment (Pierre-Gilles) [inline server/utils/constants.js:231]: Can you explain what's a counter sensor? (you can speak french ^^)

File: server/utils/constants.js [hunk 3]
```diff
@@ -247,6 +248,8 @@ const DEVICE_FEATURE_TYPES = {
     COLOR: 'color',
     TEMPERATURE: 'temperature',
     POWER: 'power',
+    EFFECT_MODE: 'effect-mode',
+    EFFECT_SPEED: 'effect-speed',
```
Comment (Pierre-Gilles) [inline server/utils/constants.js:252]: Same for the 2 previous line, can you explain what's an effect-mode/effect-speed? (you can speak french ^^)

File: server/services/tasmota/lib/setValue.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  if (prefix !== 'tasmota') {
+    throw new BadParameters(`Tasmota device external_id is invalid: "${externalId}" should starts with "tasmota:"`);
+  }
+  if (!topic || topic.length === 0) {
+    throw new BadParameters(`Tasmota device external_id is invalid: "${externalId}" have no MQTT topic`);
+  }
+
+  const gladysKey = `Gladys.${command}`;
+  const featureTemplate = FEATURE_TEMPLATES.find((template) => {
+    return template.keyMatcher.test(gladysKey);
```
Comment (Pierre-Gilles) [inline server/services/tasmota/lib/setValue.js:25]: Can you explain this block of code ? Why is there a "Gladys" key with a capital G ?

File: server/services/tasmota/lib/setValue.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  });
+
+  if (featureTemplate) {
+    const mqttValue = typeof featureTemplate.writeValue === 'function' ? featureTemplate.writeValue(value) : value;
+
+    // Send message to Tasmota topics
+    this.mqttService.device.publish(`cmnd/${topic}/${command}`, `${mqttValue}`);
```
Comment (Pierre-Gilles) [inline server/services/tasmota/lib/setValue.js:32]: Is the MQTT topic a convention of Tasmota, or is it a user-chosen one?

File: server/services/tasmota/lib/features/humidity.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      read_only: true,
+      has_feedback: false,
+      min: 0,
+      max: 100,
+      unit: DEVICE_FEATURE_UNITS.PERCENT,
+    };
+  },
+  generateExternalId: (key, fullKey) => {
+    const parts = fullKey.split('.');
+    return `${parts[1]}:Humidity`;
```
Comment (Pierre-Gilles) [inline server/services/tasmota/lib/features/humidity.js:25]: Could the "Humidity" we a capital H be in full lowercase ? Or is it a tasmota requirement?

File: server/services/tasmota/lib/features/energy.voltage.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      name: 'Voltage',
+      read_only: true,
+      has_feedback: false,
+      min: 0,
+      max: 10000,
+      unit: DEVICE_FEATURE_UNITS.VOLT,
+    };
+  },
+  generateExternalId: (key) => {
+    return `ENERGY:${key}`;
```
Comment (Pierre-Gilles) [inline server/services/tasmota/lib/features/energy.voltage.js:24]: Could the "ENERGY" be in lowercase, or is it a Tasmota requirement?

File: front/src/routes/integration/all/tasmota/device-page/index.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import { Component } from 'preact';
+import { connect } from 'unistore/preact';
+import actions from '../actions';
+import TasmotaPage from '../TasmotaPage';
+import DeviceTab from './DeviceTab';
+
+@connect('user,tasmotaDevices,housesWithRooms,getTasmotaStatus', actions)
+class TasmotaIntegration extends Component {
+  componentWillMount() {
+    this.props.getTasmotaDevices(100, 0);
```
Comment (Pierre-Gilles) [inline front/src/routes/integration/all/tasmota/device-page/index.js]: What happens if there is more than 100 Tasmota devices? 

I know it's my fault, it used to be mandatory to give a "take" parameter in a get device request, but it's no longer the case! If you want all devices, remove the take it'll give you the full list :) 

Sorry, it's really my fault here ^^

Comment (Pierre-Gilles) [reply front/src/routes/integration/all/tasmota/device-page/index.js]: See https://github.com/GladysAssistant/Gladys/blob/master/server/lib/device/device.get.js#L65

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Hi ! 

First, thanks for this really great PR ! 🙂 

There is lots of work here, and quality work, no doubt =) 

I noted a few feedbacks, most of them are just question, no change required.

There is just 2-3 comments with minor changes asked, but otherwise the rest of the code is perfect ! 

I'm looking forward to your answers to integrate this into Gladys !

**Comment by atrovato** [comment]
I have following URL on push coverage step:
https://codecov.io/github/********/Gladys/commit/a040e91670c8c62d985f8718e1ca0089575db03f

But ****** should be GladysAssistant or atrovato.

**Comment by Pierre-Gilles** [comment]
> I have following URL on push coverage step:
> https://codecov.io/github/********/Gladys/commit/a040e91670c8c62d985f8718e1ca0089575db03f
> 
> But ****** should be GladysAssistant or atrovato.

When I think about it I'm not sure this can work... Does codecov allow a forked repo to push data on my Codecov account ? I think it would be a security issue and I don't know if they allow that

No worries, I can push your branch on Gladys main repo before merging to have a full build in any case.

**Comment by Pierre-Gilles** [comment]
In fact, it works on other PR from fork! 

Example:

https://codecov.io/gh/GladysAssistant/Gladys/pull/675

Maybe you are doing something differently? Did it work in the past for others PR you did?

File: server/utils/constants.js (hunk 2 above)
Comment (atrovato) [reply server/utils/constants.js:231]: C'est un compteur simple, capable d'incrémenter sa valeur.
Sur le site de Tasmota, ils parlent principalement du nombre de "pulse" émis sur un PIN de l'esp.

File: server/utils/constants.js (hunk 3 above)
Comment (atrovato) [reply server/utils/constants.js:252]: Effect mode : stroboscope, dégradé de lumière...
Effect speed: vitesse du flash du stroboscope, vitesse du changement de couleur

File: server/services/tasmota/lib/setValue.js (hunk 5 above)
Comment (atrovato) [reply server/services/tasmota/lib/setValue.js:32]: It's a Tasmota convention:
MQTT
https://tasmota.github.io/docs/#/MQTT?id=commands-over-mqtt 
Commands
https://tasmota.github.io/docs/#/Commands

File: server/services/tasmota/lib/setValue.js (hunk 4 above)
Comment (atrovato) [reply server/services/tasmota/lib/setValue.js:25]: It's a shortcut to keep the same rule as MQTT message, but with `Gladys` key work.
The `keyMatcher` should match `StatusSTS` form Tasmota or `Gladys` for the `setValue` case.

File: server/services/tasmota/lib/features/humidity.js (hunk 6 above)
Comment (atrovato) [reply server/services/tasmota/lib/features/humidity.js:25]: To make easier the way to match Tasmota and Gladys, the feature external id contains the exact key sent by the Tasmota firmware.
So it needs to  match case.

File: server/services/tasmota/lib/features/energy.voltage.js (hunk 7 above)
Comment (atrovato) [reply server/services/tasmota/lib/features/energy.voltage.js:24]: Same as Humidy.

> To make easier the way to match Tasmota and Gladys, the feature external id contains the exact key sent by the Tasmota firmware.
So it needs to match case.

File: front/src/routes/integration/all/tasmota/device-page/index.js (hunk 8 above)
Comment (atrovato) [reply front/src/routes/integration/all/tasmota/device-page/index.js]: Ok, we'll need to check other services.

File: front/src/components/boxs/device-in-room/device-features/BinaryDeviceFeature.jsx (hunk 1 above)
Comment (atrovato) [reply front/src/components/boxs/device-in-room/device-features/BinaryDeviceFeature.jsx]: Ok, I did it to please to testers, I'll remove it.

**Comment by atrovato** [comment]
Ok, for test coverage, I unfollow my own GitHub repository, URL looks good for now.

**Review by Pierre-Gilles — APPROVED** [review]
